logs: Update decision logger to support masking

tsandall opened the pull request:
These changes enhance the in-built decision logger to support masking
of input and result fields for cases where sensitive information is
passed to OPA (or generated by the policy.)

Also, fix rego package to avoid panic-ing on programmatically created
queriest that lack Location information.

Fixes #779

Signed-off-by: Torin Sandall <torinsandall@gmail.com>

tsandall commented:
cc @koponen-styra

tsandall commented:
cc @arcurtis

patrick-east commented on plugins/logs/plugin.go:
```diff
@@ -385,6 +415,37 @@ func (p *Plugin) bufferChunk(buffer *logBuffer, bs []byte) {
+		return nil
+	}
+
+	ptrs, err := resultValueToPtrs(rs[0].Expressions[0].Value)
```
Is it safe to assume `Expressions[0]` will exist?

tsandall replied:
Yes, it's safe because the query contains exactly one expression and this is how the value of the expression is accessed. Alternatively we could construct an expression like `data.system.log.mask = x` and then lookup the x binding but the same logic would apply.

patrick-east replied:
Ahh ok, that makes sense. I missed that `parsePathToRef` wouldn't allow someone to configure an empty query or something.

patrick-east commented on plugins/logs/plugin.go:
```diff
@@ -385,6 +415,37 @@ func (p *Plugin) bufferChunk(buffer *logBuffer, bs []byte) {
 	}
 }
 
+func (p *Plugin) maskEvent(ctx context.Context, event *EventV1) error {
```
Is it possible to check first if the policy has been defined? Could maybe avoid some of the extra work being done here if there isn't one set by the deployer.

tsandall replied:
I think if this becomes an issue we could add a setting to disable masking. My intuition is that most users will need masking if they have decision logging enabled.

koponen-styra commented on plugins/logs/plugin.go:
```diff
@@ -385,6 +415,37 @@ func (p *Plugin) bufferChunk(buffer *logBuffer, bs []byte) {
+
+	query := ast.NewBody(ast.NewExpr(ast.NewTerm(p.config.maskDecisionRef)))
+
+	r := rego.New(
```
This is running policy for each event, right? Is that going to be an issue (CPU)?

tsandall commented:
@patrick-east @koponen-styra I've updated the PR to lazily prepare the masking query. This reduces the masking query latency by ~50% (down to ~150 microseconds on my machine in the no-op case). I've also updated the test suite to include a benchmark for the masking operation.
